RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitPhase2WeightsAlgoGPU.cu

```diff
@@ -7,18 +7,20 @@
 namespace ecal {
   namespace weights {
 
-    void entryPoint(ecal::DigisCollection<calo::common::DevStoragePolicy> const& digis,
-                    EventOutputDataGPU& eventOutputGPU,
-                    cms::cuda::device::unique_ptr<double[]>& weights_d,
-                    cudaStream_t cudaStream) {
-      unsigned int totalChannels = digis.size;
+    void phase2Weights(ecal::DigisCollection<calo::common::DevStoragePolicy> const& digis,
+                       EventOutputDataGPU& eventOutputGPU,
+                       cms::cuda::device::unique_ptr<double[]>& weights_d,
+                       cudaStream_t cudaStream) {
+      unsigned int const totalChannels = digis.size;
       // 64 threads per block best occupancy from Nsight compute profiler
-      unsigned int nchannels_per_block = 64;
-      unsigned int threads_1d = nchannels_per_block;
-      unsigned int blocks_1d = (totalChannels / threads_1d) + 1;
+      unsigned int const nchannels_per_block = 64;
+      unsigned int const threads_1d = nchannels_per_block;
+      unsigned int const blocks_1d = (totalChannels + threads_1d - 1) / threads_1d;
+      unsigned int const n_iter = (totalChannels + threads_1d * blocks_1d - 1) / (threads_1d * blocks_1d);
```

Comment:
I don't understand the logic here :-(

If I disregard the rounding, the constants become
```c++
unsigned int const nchannels_per_block = 64;
unsigned int const threads_1d = nchannels_per_block;
unsigned int const blocks_1d = totalChannels  / threads_1d;
unsigned int const n_iter = totalChannels / (threads_1d * blocks_1d);
```
and then
```c++
unsigned int const n_iter = totalChannels / (threads_1d * totalChannels  / threads_1d);
```
so
```
unsigned int const n_iter = 1;
```
?

Reply:
The `n_iter` variable was introduced to allow any value of `blocks_1d` to be used in the future. In this case `blocks_1d` is chosen such that `n_iter` is equal to 1.

Reply:
I guessed so, but with the current definition of the `blocks_1d` and `n_iter` variables, `n_iter` will always be 1 (or possibly 2 because of rounding), no matter the value of `threads_1d`.

Reply:
Could you give an example of how one would set different values for `threads_1d` and `blocks_1d` ?

Reply:
At the moment it is just done by changing the lines:
`unsigned int const nchannels_per_block = 64;
unsigned int const blocks_1d = totalChannels  / threads_1d;`

Reply:
Yes we needed it for the dynamic shared memory since it is per block.

Reply:
> At the moment it is just done by changing the lines: `unsigned int const nchannels_per_block = 64;` and ` unsigned int const blocks_1d = totalChannels / threads_1d;`

Can you elaborate how you would run the kernel with 1 block and 1 thread ?

> Yes we needed it for the dynamic shared memory since it is per block.

And what should the memory allocation be, in that case ?

Reply:
OK, I'm starting to follow...

What is the difference between `nchannels_per_block` and `threads_1d` ?
What I mean is, if we run `kernel<<<1, 1>>>`, would `nchannels_per_block` be `1` ? or still `64` ?

Reply:
There is no difference, we could remove `nchannels_per_block` and replace instances of it with `threads_1d`, it is a variable which is used by EcalUncalibRecHitMultiFitAlgoGPU.cu where the two are different but here they are the same.